### PR TITLE
New spi modules

### PR DESCRIPTION
**Pull request by bradleyharden**
This is a pretty big addition. I'm marking it as WIP mostly because I need #401 to be merged first. But I also think it might take some time in review.

The best way to get started is probably the documentation. Is there some way for me to render it for reviewers? Or do reviewers need to download it and render it locally?

@twitchyliquid64, you have talked through some of the SPI stuff with me before. @blm768, this is an extension of the GPIO and SERCOM pad work from before. @jbeaurivage, I think there's a lot of overlap in approach between this and your DMAC stuff. Although safety is less prominent of an issue for this.

I also added a few dependencies. I'm not sure how averse you are to that. If you have any specific objections, there's probably a workaround.

### Commit message

Move declarations of the `Sercom` types from the `pads` module to the base `sercom` module. Make them type aliases of the PAC SERCOM types. Add a `Sercom` trait method to enable the APB clock.

Create new SPI modules for both `thumbv6m` and `thumbv7em`. The new modules are more complete, more configurable and safer than the existing modules. They make use of the `v2` `gpio` and `sercom::pads` modules.

Also add an `SpiFuture` type that can execute SPI transactions autonomously, using interrupts.

### CODE REVIEW

File: hal/Cargo.toml
Comment (Sympatron) [inline hal/Cargo.toml:31]: `generic-array` re-exports `typenum`. Maybe it would be better to use that to avoid accidental version mismatches in the future? Not a major problem in my opinion, just a thought.

File: hal/src/common/sercom/v2/spi_future.rs
Comment (Sympatron) [inline hal/src/common/sercom/v2/spi_future.rs:434]: Shouldn't this read and all writes to `self.rcvd` be volatile?

File: hal/Cargo.toml
Comment (bradleyharden) [reply hal/Cargo.toml:31]: Yeah, that's a good idea. I don't really need a specific version of either. I'll do that.

File: hal/src/common/sercom/v2/spi_future.rs
Comment (bradleyharden) [reply hal/src/common/sercom/v2/spi_future.rs:434]: Hmm. I don't think so. But maybe there's something subtle I'm missing.

I think you're suggesting that multiple reads of `self.rcvd` by repeated calls to `poll` would be optimized away, so you would never see updates to `self.rcvd`. Is that right?

If you're using accessing `SpiFuture` from interrupts at different priority levels, then it would need to be wrapped in a `Mutex` of some type. In RTIC, you would `poll` from a closure passed to the `Mutex`. There's no `Mutex` type in `core`, so I'm not sure what you would do there. But any solution would probably look like `std::sync::Mutex`, where locking would create a new `MutexGuard`. In either case, it wouldn't make sense to `poll` more than once, since the `Mutex` ensures your access can't be preempted.

If you're accessing `SpiFuture` from interrupts at the same priority level, then they can't preempt each other to begin with. So once again, it doesn't make sense to `poll` more than once.

Is there a scenario I'm missing? Or something I'm misunderstanding?

**Comment by bradleyharden** [comment]
I just realized something I would like to add. Right now, the SPI transaction `Length` (1-255 bytes) for SAMD51 & SAME5x chips **must** be known at compile time. I think that works well for most cases, but there are probably some cases where you would want to set the length dynamically.

I think I will add an extra, marker type for the `Length` trait that indicates a run-time dynamic length. I think that should be easy to add.

File: hal/src/common/thumbv6m/sercom/v2/spi.rs
Comment (twitchyliquid64) [inline hal/src/common/thumbv6m/sercom/v2/spi.rs:1127]: `Mode`, `Phase`, `Polarity` are part of the API, could we re-export so peeps dont have to throw `use atsamd_hal::hal::spi::{Mode, Phase, Polarity};` everywhere?

File: hal/src/common/thumbv6m/sercom/v2/spi.rs
Comment (twitchyliquid64) [inline hal/src/common/thumbv6m/sercom/v2/spi.rs:341]: Encapsulates

File: hal/src/common/thumbv6m/sercom/v2/spi.rs
Comment (twitchyliquid64) [inline hal/src/common/thumbv6m/sercom/v2/spi.rs:378]: Maybe add a comment that `data_in` corresponds to miso when SPI is in master mode? or on the `data_in` docs?

File: hal/src/common/thumbv6m/sercom/v2/spi.rs
Comment (twitchyliquid64) [inline hal/src/common/thumbv6m/sercom/v2/spi.rs:381]: It might be useful as well to add a note that you need to provide the pins as `v2` GPIO types, to avoid needing verbose type declarations.

File: hal/src/common/thumbv6m/sercom/v2/spi.rs
Comment (twitchyliquid64) [inline hal/src/common/thumbv6m/sercom/v2/spi.rs:1099]: It might be useful to specify the default here?

**Review by twitchyliquid64 — APPROVED** [review]
Looks great thanks! Just a few minor comments.

Heres a patch for the wio_terminal display system that switches it to the new API, feel free to throw it in too.

```
diff --git a/boards/wio_terminal/src/display.rs b/boards/wio_terminal/src/display.rs
index 1bdb32c21..5b576f9fa 100644
--- a/boards/wio_terminal/src/display.rs
+++ b/boards/wio_terminal/src/display.rs
@@ -1,14 +1,17 @@
 use atsamd_hal::clock::GenericClockController;
 use atsamd_hal::gpio::*;
 use atsamd_hal::hal::blocking::delay::DelayMs;
-use atsamd_hal::hal::spi;
 use atsamd_hal::prelude::*;
-use atsamd_hal::sercom::{PadPin, SPIMaster7, Sercom7Pad1, Sercom7Pad2, Sercom7Pad3};
 use atsamd_hal::target_device::{MCLK, SERCOM7};
 use atsamd_hal::time::Hertz;
 use display_interface_spi::SPIInterface;
 use ili9341::{Ili9341, Orientation};

+use atsamd_hal::hal::spi::{Mode, Phase, Polarity};
+use atsamd_hal::sercom::v2::pads::{IoSet4, Pad, Pad1, Pad3};
+use atsamd_hal::sercom::v2::{spi, Sercom7};
+use atsamd_hal::typelevel::NoneT;
+
 /// ILI9341 LCD display pins (uses `SERCOM7`)
 pub struct Display {
     /// LCD MISO pin
@@ -36,7 +39,17 @@ pub struct Display {
 /// Type alias for the ILI9341 LCD display.
 pub type LCD = Ili9341<
     SPIInterface<
-        SPIMaster7<Sercom7Pad2<Pb18<PfD>>, Sercom7Pad3<Pb19<PfD>>, Sercom7Pad1<Pb20<PfD>>>,
+        spi::Spi<
+            spi::Config<
+                spi::Pads<
+                    SERCOM7,
+                    IoSet4,
+                    NoneT,
+                    Pad<SERCOM7, Pad3, IoSet4>,
+                    Pad<SERCOM7, Pad1, IoSet4>,
+                >,
+            >,
+        >,
         Pc6<Output<PushPull>>,
         Pb21<Output<PushPull>>,
     >,
@@ -58,23 +71,27 @@ impl Display {
         clock_speed: F,
         delay: &mut D,
     ) -> Result<(LCD, Pc5<Output<PushPull>>), ()> {
-        // Initialize the SPI peripherial on the configured pins, using SERCOM7.
+        let sclk: atsamd_hal::gpio::v2::Pin<_, _> = self.sck.into();
+        let mosi: atsamd_hal::gpio::v2::Pin<_, _> = self.mosi.into();
+
+        let pads = spi::Pads::<Sercom7, IoSet4>::new()
+            .sclk(sclk)
+            .data_out::<Pad3, _>(mosi);
+
+        // Wire the clock for SERCOM7 to the system clock.
         let gclk0 = clocks.gclk0();
-        let spi = SPIMaster7::new(
-            &clocks.sercom7_core(&gclk0).ok_or(())?,
-            clock_speed,
-            spi::Mode {
-                phase: spi::Phase::CaptureOnFirstTransition,
-                polarity: spi::Polarity::IdleLow,
-            },
-            sercom7,
-            mclk,
-            (
-                self.miso.into_pad(port),
-                self.mosi.into_pad(port),
-                self.sck.into_pad(port),
-            ),
-        );
+        clocks.sercom7_core(&gclk0).ok_or(())?;
+
+        // Initialize the SPI peripherial on the configured pads, using SERCOM7.
+        let spi = spi::Config::new(mclk, sercom7, pads, gclk0)
+            .baud(clock_speed)
+            .length::<spi::lengths::U1>()
+            .msb_first(true)
+            .spi_mode(Mode {
+                phase: Phase::CaptureOnFirstTransition,
+                polarity: Polarity::IdleLow,
+            })
+            .enable();

         // Configure the chip select, data/command, and reset pins as push-pull outputs.
         let cs = self.cs.into_push_pull_output(port);
```

File: hal/src/common/thumbv6m/sercom/v2/spi.rs
Comment (bradleyharden) [reply hal/src/common/thumbv6m/sercom/v2/spi.rs:1127]: Good idea

File: hal/src/common/thumbv6m/sercom/v2/spi.rs
Comment (bradleyharden) [reply hal/src/common/thumbv6m/sercom/v2/spi.rs:341]: I meant it in the imperative mood. But I may not have been consistent across  the docs. I'll check it and make sure

File: hal/src/common/thumbv6m/sercom/v2/spi.rs
Comment (bradleyharden) [reply hal/src/common/thumbv6m/sercom/v2/spi.rs:378]: Good call. Will do

File: hal/src/common/thumbv6m/sercom/v2/spi.rs
Comment (bradleyharden) [reply hal/src/common/thumbv6m/sercom/v2/spi.rs:1099]: Good catch

File: hal/src/common/thumbv6m/sercom/v2/spi.rs
Comment (bradleyharden) [reply hal/src/common/thumbv6m/sercom/v2/spi.rs:381]: I thought I designed it to work with either pin type. Is that not working for you? See my comment on Matrix

Comment (twitchyliquid64) [reply hal/src/common/thumbv6m/sercom/v2/spi.rs:381]: Nop, due to ambiguity in type inference :/ More details in my comment on matrix

Comment (twitchyliquid64) [reply hal/src/common/thumbv6m/sercom/v2/spi.rs:381]: With into:

```
cannot infer type for struct `Pad<SERCOM7, Pad1, IoSet4>`
```

Without:
```
the trait `From<atsamd_hal::gpio::Pin<PB20, atsamd_hal::gpio::v2::Input<atsamd_hal::gpio::v2::Floating>>>` is not implemented for `Pad<SERCOM7, Pad1, IoSet4>
```

File: hal/src/common/thumbv6m/sercom/v2/spi.rs
Comment (bradleyharden) [reply hal/src/common/thumbv6m/sercom/v2/spi.rs:1099]: Wait, what do you mean by default? Do you mean the default `Config` returned by `new`? Looks like I forgot to document the default `CharSize` there. But there's no default for the `char_size` function. You have to provide the desired type.

File: hal/src/common/thumbv6m/sercom/v2/spi.rs
Comment (bradleyharden) [reply hal/src/common/thumbv6m/sercom/v2/spi.rs:381]: I should have a fix for this.

File: hal/src/common/sercom/v2/spi_future.rs
Comment (Sympatron) [reply hal/src/common/sercom/v2/spi_future.rs:434]: You obviously thought this through a lot more than I have. I am pretty sure that you would need volatile for similar code in C, which I am more used to. I forget that Rust's ownership/safety model prevents these kinds of mistakes.

File: hal/src/common/thumbv6m/sercom/v2/spi.rs
Comment (bradleyharden) [reply hal/src/common/thumbv6m/sercom/v2/spi.rs:1099]: Right now, the default `CharSize` is documented in the module-level docs and the docs for `Config`, but it is not documented in the `Config::new` function. I added it there as well. Is that sufficient?

File: hal/src/common/thumbv6m/sercom/v2/spi.rs
Comment (bradleyharden) [reply hal/src/common/thumbv6m/sercom/v2/spi.rs:1127]: One problem with this. I define my own `Mode` trait in this module. I think I will skip re-exporting the embedded HAL `Mode`. If you're using the embedded HAL `Mode`, you're probably using one of the constants, e.g. `MODE_0`.

File: hal/src/common/sercom/v2/spi_future.rs
Comment (bradleyharden) [reply hal/src/common/sercom/v2/spi_future.rs:434]: Just because I've thought about it doesn't mean it's correct =D

I'm relying on the assumption that using only safe code should do what I want/expect to happen. I think that's right in this case, but I'm not 100% sure.

File: hal/src/common/thumbv6m/sercom/v2/spi.rs
Comment (bradleyharden) [reply hal/src/common/thumbv6m/sercom/v2/spi.rs:341]: Until you prompted me here, I hadn't actually thought that much about how I write documentation. It looks like I generally stick to two approaches. Either I lead with a descriptive noun phrase and then describe what it does, e.g.

> Type-level `enum` representing the SPI transaction length, in bytes

Or I lead with an imperative that describes what it does, e.g.

> Control the `DIPO` field as a function of the [`PadNum`] type

I think I use the latter when I don't have any meaningful additions to the noun. For example,

> Trait that controls the `DIPO` field as a function of the [`PadNum`] type

is worse, because "Trait that" adds no real value. We already know it's a trait. Instead, I opt for the imperative, because it's usually terse. I think the same reasoning applies for using the imperative in Git commit messages, which I follow.

File: hal/src/common/thumbv6m/sercom/v2/spi.rs
Comment (bradleyharden) [reply hal/src/common/thumbv6m/sercom/v2/spi.rs:1127]: Actually, I'll just export the `Mode` constants instead.

**Comment by bradleyharden** [comment]
Just updated with the suggested changes. Please merge #406 before this.

If no one else has any comments or feedback, I guess this could be ready to merge.

File: hal/src/common/thumbv6m/sercom/v2/spi.rs
Comment (bradleyharden) [reply hal/src/common/thumbv6m/sercom/v2/spi.rs:393]: I can confirm this works with `v1` pins with your Wio Terminal diff

**Comment by bradleyharden** [comment]
Don't merge just yet. I realized I forgot some pieces related to the new run-time dynamic length feature.

File: hal/src/common/thumbv6m/sercom/v2/spi.rs
Comment (twitchyliquid64) [reply hal/src/common/thumbv6m/sercom/v2/spi.rs:353]: I dont feel strongly about this :joy: I mentioned it as I'm used to the Go style guide, which suggests all type-level doc strings should be full sentences (ie: subject + verb + what) even if it reads verbose.

Your logic makes sense

File: hal/src/common/thumbv6m/sercom/v2/spi.rs
Comment (twitchyliquid64) [reply hal/src/common/thumbv6m/sercom/v2/spi.rs:1111]: SGTM

**Review by twitchyliquid64 — APPROVED** [review]
Thanks for this! the new trait-based API is a big step forward.

Will merge once green!

File: hal/src/common/thumbv6m/sercom/v2/spi.rs
Comment (bradleyharden) [reply hal/src/common/thumbv6m/sercom/v2/spi.rs:353]: Interesting, I'll have to read up on it. It probably sounds crazy, but I really like to have some justification for every decision, no matter how ultimately arbitrary it is. Ultimate bikeshedding I guess :p 

I always like to read arguments for/against certain styles and that sort of thing. I find it interesting how people reason about stuff.

File: hal/src/common/sercom/v2/spi_future.rs
Comment (Sympatron) [reply hal/src/common/sercom/v2/spi_future.rs:514]: I still have to learn to trust the Rust compiler more than I can trust any C compiler. Years of writing C taught me to be very careful and it's hard to unlearn. :upside_down_face:

**Pull request merged**
